src/engine/sparqlExpressions/AggregateExpression.h

```diff
@@ -43,39 +41,60 @@ inline auto getUniqueElements = []<typename OperandGenerator>(
+
+// Class for a SPARQL expression that aggregates a given set of values to a
+// single value using `AggregateOperation`, and then applies `FinalOperation`.
+//
+// NOTE: The `FinalOperation` is typically the `identity` from above. One
+// exception is the `AvgExpression`, where the `FinalOperation` divides the
+// aggregated value (sum) by the number of elements.
+template <typename AggregateOperation,
+          typename FinalOperation = decltype(identity)>
 class AggregateExpression : public SparqlExpression {
  public:
-  // __________________________________________________________________________
+  // Create an aggregate expression, where `child` is the expression to be
+  // aggregated.
+  //
+  // For example, for `SUM(?x + 5)`, `child` is the expression for `?x + 5`,
+  // `distinct` is `false`, and `aggregateOp` is the operation for computing
+  // the sum.
   AggregateExpression(bool distinct, Ptr&& child,
                       AggregateOperation aggregateOp = AggregateOperation{});
 
-  // __________________________________________________________________________
+  // Evaluate this aggregate expression.
   ExpressionResult evaluate(EvaluationContext* context) const override;
 
-  // _________________________________________________________________________
+  // Each aggregate expression has to say what its result for an empty group is
+  // (needed only for implicit GROUP BYs).
+  virtual ValueId resultForEmptyGroup() const = 0;
+
+  // Get the unaggregated variables of the expression (can be empty).
   vector<Variable> getUnaggregatedVariables() override;
 
-  // _________________________________________________________________________
+  // Yes, this is an aggregate expression.
   bool isAggregate() const override { return true; }
 
-  // __________________________________________________________________________
+  // Get the cache key for this expression.
   [[nodiscard]] string getCacheKey(
       const VariableToColumnMap& varColMap) const override;
 
-  // __________________________________________________________________________
+  // This is only used for aggregate expressions, and then it's exactly the
+  // value of the `distinct` parameter in the constructor.
   bool isDistinct() const override { return _distinct; }
 
-  // __________________________________________________________________________
+  // Needed for the pattern trick, see `SparqlExpression.h`.
   [[nodiscard]] std::optional<SparqlExpressionPimpl::VariableAndDistinctness>
   getVariableForCount() const override;
 
-  // This is the visitor for the `evaluateAggregateExpression` function below.
-  // It works on a `SingleExpressionResult` rather than on the
-  // `ExpressionResult` variant.
+  // Evaluate a `SingleExpressionResult` (that is, one of the possible
+  // `ExpressionResult` variants). Used in the `evaluate` function.
+  //
+  // TODO: Why is this a lambda and not a normal member function? It's rather
+  // long and complex.
```

Comment:
It cannot be a member function because it is templated on its argument, but has to be passed to another function ( std::visit in this case). 
You could however reimplement this as a (non-templated) class with a templated call operator (the exact equivalent of the lambda), then the definition can land in the cpp file.